### ISSUE TITLE
508-defect-3: Focus should be sent to a background-only alert message indicating an update has been saved

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#53446]
## Point of contact

joshkimux

## Severity level

2, Serious. Should be fixed in 1-2 sprints post-launch.

## Details

Two interconnected issues exist when editing contact information:

1. Focus isn't managed upon returning to the contact information page after clicking "Cancel" or "Save" post editing one of the user's contacts and 
2. no aural (or immediate visual) confirmation is communicated to the user that the change has been saved

## Reproduction steps

1. "On the contact information page, click "edit" on any of the options.

2. Turn on your screen reader of choice.

3. Click "save" which triggers an aural "saving" message

4. Confirm you are returned to the contact information page, but hear no confirmation message and see no immediate feedback that it was saved or changed.

<details><summary> Old User Story Content</summary>

## Proposed solution or next steps

<del>

This one is a bit tricky as we're trying to communicate both a change of state (you've successfully updated something) and create consistent and predictable navigation (you're returned back to where you started). 

This works great when adding issues, as a literal new checkbox is created with the user's information which can take focus-- but in this case, it's not as clear in this context as static text is being updated. 

@andaleliz  on the profile team has conducted research on this. They may have evidence-backed guidance for us to follow here. We can likely solve this by either introducing one more page for confirmation pages (we've updated your number to this) OR add a polite visual/aural alert upon returning to the introduction page "We've updated your phone number in your profile."

## Update ⚠️ Non-design fixes to address major issues until post-launch fixes can be made informed by usability testing
After meeting with @eileen-coforma , we decided on implementing non-design fixes as a temporary solution until post-launch fixes can be made that are informed by usability testing with screen reader users.

We recommend the following short-term solutions to resolve this ticket:
- Post editing or cancellation, focus must return to the "edit" link that triggered the flow (for circular focus)
  - Fixing this _should_ will resolve the first issue of this ticket
  - Pending platform's guidance, the second part of this issue on aural/visual feedback that a save has been made may range between a defect 3-4
- To address the second part of the issue, we could consider the following non-design change to reduce the defect level (or fully address it)
  - Add an sr-only `aria-live` region to announce the change has been saved e.g. "Your mobile phone number has been updated to 703-132-2341"
- In the future, more holistic changes (pending usability testing) may include:
  - Using Liz's update pattern from profile which tested favorably. This may need some tweaks as the current design already has a left border.
  - ??
</del>
</details>

## Update ⚠️ , post meeting with the platform team
- The platform team noted that missing a visual alert indicating a change has been saved would warrant a `defect-level-2`. Low vision users and users with cognitive disabilities may be unable to detect a change has been saved without a visual signifier.
- @allison0034 recommended using the VA background-only alert
- @briandeconinck recommended
  - messaging should be specific "[name of contact method] updated" e.g. "Home phone number updated"
  - focus should be sent to the background-only alert upon _save_ (allowing users to navigate downwards aurally to the new number/email/address and then again to the edit link)
  - focus should be sent to the `edit` link on `cancel` or no changes

### Success criteria

- [x] Focus returns to the edit link that triggered the flow upon "cancel" for all contact methods:
  - [x] Mobile phone number
  - [x] Home phone number
  - [x] Email
  - [x] Address 
- [x] A _success_  background alert is provided between the unique contact heading and the string below it
  - [x] The alert's content reads as "[name of contact method] updated"
  - [x] Focus is sent to the background alert post saving on the prior page
- [ ] `role="alert"` is added to the background-only alert to announce its contents for JAWs and VoiceOver/iOS users
- [ ] Error messaging is provided if the user attempts to click "save" without making any changes

#### VoiceOver/Safari iOS
Works like butter!

#### NVDA/Chrome
Works fantastic!

#### VoiceOver/Safari Macbook 

https://user-images.githubusercontent.com/14154792/222269969-ff7df103-321d-484c-adb7-262539760d00.mov

- Focus sometimes is lost when returning to the contact information page after a save
- "Save" message sometimes cancels out `h3` "Edit..." on a screen reader

#### JAWs
- JAWs does not announce "home phone number updated" upon making a change ("saving changes, file a supplemental claims, main region..." is announced instead, which likely reads over focus)
- Using an `aria-live` polite region _may_ address this, at the expense of more verbosity for the other screen readers. 

## References, articles, or WCAG support

- [Focus Order: Understanding SC 2.4.3](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-order.html)

## Type of issue

- [X] Screenreader
- [X] Keyboard
- [X] Focus
- [ ] Headings
- [ ] Color
- [ ] Zoom
- [ ] Semantics
- [ ] Axe-core
- [ ] Something else

**Comment by andaleliz** [comment]
@joshkimux thanks for tagging me in on this. I _think_ based on the supplemental claims label this is NOT a profile issue, but please lmk if I'm wrong about that.

**Comment by joshkimux** [comment]
@andaleliz not a profile issue at all! 🙇 Just pointing to the recent research that you did on this that we could learn from!

**Comment by Mottie** [comment]
Adding a needs design tag to this. I've spoken to @eileen-coforma about adding some indicator that shows an entry was either updated or cancelled

**Comment by joshkimux** [comment]
**Update**

Post meeting with @eileen-coforma , we decided to do the following:
- determine a non-design fix to this issue as a temporary solution to keep us on track to launch
- a non-design fix should drop this issue to a defect-3 or 4, provided it can be worked on post-launch

We agreed on this route as we knew that determining a solution would be difficult in the absence of usability testing with screen reader users.

**Comment by joshkimux** [comment]
@Mottie if we address the focus management bit of this ticket in the updated success criteria:

> Focus returns to the edit link that triggered the flow upon "save" or "cancel" for all contact methods:
>  Mobile phone number
>  Home phone number
>  Email
>  Address

We should be able to knock this down to a defect 3 or 4 pending platform review. I'll be taking this in to office hours on Wednesday on behalf of the team.

@saderagsdale for awareness ✨

**Comment by saderagsdale** [comment]
@Mottie Does this ticket still need design attention?

**Comment by Mottie** [comment]
Yes. But I have work-in-progress to knock this down from a 508-defect-2, so it won't be a launch blocker.

**Comment by saderagsdale** [comment]
@joshkimux let us know what you hear during office hours

**Comment by joshkimux** [comment]
Platform recommendation:
- [Use existing background only alert](https://design.va.gov/components/alert#examples---background-color-only)
- With messaging "[x] updated"

<img width="666" alt="Screenshot 2023-02-22 at 11 49 26 AM" src="https://user-images.githubusercontent.com/14154792/220698891-77edea16-2347-436c-99f1-52bfb596fc0e.png">

**Comment by joshkimux** [comment]
@saderagsdale and @eileen-coforma updated to match platform's feedback ❤️ 🖖

**Comment by saderagsdale** [comment]
Hid the content you scratched out.

**Comment by Mottie** [comment]
What would it show if they canceled? No message at all? Or would it be a "Home phone number change was canceled", on a yellow or blue background?

**Comment by joshkimux** [comment]
@Mottie no message at all! Focus would simply return to the edit link.

**Comment by Mottie** [comment]
Work merged in and is now available in staging for testing.

**Comment by joshkimux** [comment]
@Mottie so freaking good!! I did catch a couple of bugs (updated original ticket), namely:
- Focus sometimes is lost when returning to the contact information page after a save
- "Save" message sometimes cancels out h3 "Edit..." on a screen reader
- Error messaging is provided if the user attempts to click "save" without making any changes

I'm not too confident on how we might fix this. My non-engineering instinct (which is likely wrong) is that this may be prickly matter with React. If we focus() too early, the render may be wiping out out our focus. 

If that's the case, setTimeout _could_ help by moving the focus call to the next call stack after render completes.

@jaredcunha (who is an actual engineer) may have a better symptom/diagnosis!

**Comment by Mottie** [comment]
Oh, thanks! I'll take a look!

**Comment by Mottie** [comment]
Hmmm, this is a tough one. I can duplicate it staging occasionally, but it is exceedingly difficult to test this in my local dev environment. I'll try again, but finding the source of this issue may take some time 😿

**Comment by joshkimux** [comment]
@Mottie , tested on Chrome/NVDA and it's working like a dream there. Not sure why it's bugging up on VoiceOver. Need to test on JAWs.

**Comment by joshkimux** [comment]
Another update, after testing...
- VoiceOver iOS ✔️ 
- NVDA Chrome ✔️ 
- Jaws 😐 
- VoiceOver Desktop 😢 

I'm going to bump this down to a defect-level 3 as:
- Focus still works on JAWs, it just doesn't aurally announce the update. We may be able to fix this using a polite `aria-live` region on top of focus management (but this may make it more verbose on the other screen reader pairings)
- Safari/VO on desktop, from what we know, is likely the least used screenreader combination Veterans are likely to use

**Comment by joshkimux** [comment]
@briandeconinck when you return, I'd love to get your thoughts on where this would sit in terms of severity. It performs great on VO/iOS and NVDA/Chrome, but doesn't get the desired results on JAWs and VoiceOver for desktop. 

Ideally, the message would still announce on JAWs given many screen reader Veterans are JAWs users (hence keeping this at a defect-3)-- but I'm not too confident on possible solutions.

**Comment by briandeconinck** [comment]
@joshkimux I think I agree on labeling this a `defect-3`. That VO behavior is so strange though... I wonder if adding `role="alert"` to the success message buys you anything?

**Comment by joshkimux** [comment]
`role="alert"` may do the trick in this instance, but I'm a little worried it might be a bit too aggressive. 

> Warning: Because of its intrusive nature, the alert role must be used sparingly and only in situations where the user's immediate attention is required.
- [MDN webdocs, ARIA: alert role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/alert_role)

I don't think `status` would be a better alternative though, as it would theoretically add to the verbosity (focus + status announcement on VO and NVDA). `alert` at least would drown everything else out.

> Do not give focus to the status when its content updates. Live regions are meant to inform users of dynamic updates that have occurred in other areas of the current web page, but which do not necessitate interrupting the user's current activity with a change in context. If a situation requires that focus needs to be moved, then using a status, or other live region, are likely not appropriate.
- [MDN webdocs, ARIA: status role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/status_role)

Which leads me to the question of, which is worse?
- An intrusive alert role OR
- Inconsistent announcements on VO/Macbook and missed announcement on JAWs

I think after mapping it out, I agree with @briandeconinck 's assessment, provided we document our decision making behind it. Follow up question would be:
- Should we change how we manage focus if we were to use the `alert` role? e.g. since it will be announced regardless, should we send focus... 
  - back to `edit` (make it circular, but this is _after_ the updated number, so the user will need to remember to step backwards to hear the update)
  - to the phone number (return to the updated number, but this is a static field that doesn't typically receive focus)
  - to the alert (which is redundant to the `alert` role, but allows the user to progress to the updated phone number next)

**Comment by joshkimux** [comment]
@Mottie update! I talked to Brian about this, and we've decided to:
- use the `alert` role (and take the L on it being noisy to take the W on it consistently announcing)
- keep focus on the alert itself to maintain the navigation order

**Comment by Mottie** [comment]
Hot potato back at ya... ready for review in staging

**Comment by joshkimux** [comment]
https://user-images.githubusercontent.com/14154792/224691508-0763cb14-8c02-4348-b5c6-a3d32e4a8d36.mov

Tested on JAWS/Chrome and Macbook/VO again! Some good news (and bad news 😢 )
- It announces consistently on Macbook/VO 🥳 
- ...but it still doesn't announce on JAWS/Chrome and
- there's a slight bug where the alert reads out twice when reaching the edit link 😢 

@Mottie I'm going to take this to a11y office hours for further investigation. I can't seem to figure out what the issue is here, especially on JAWS/Chrome which _should_ be the least likely to have an issue.

**Comment by joshkimux** [comment]
@Mottie Update! Met with Angela to noodle on this, and we came up with two possible ways of moving forward. Would love your thoughts on which makes more sense / is less difficult to implement. A note here, this is still a `defect-3`, so please deprioritize this in your list of things to tackle!

### Option A (preferred): Remove the "saving" announcement when clicking "save"
In theory, this would remove unnecessary verbosity that may be drowning out the "Your x has been updated" announcement by `role="alert"` / focus management.

### Option B: Delay the `aria-live` announcement  
Instead of using an immediate aria `role="alert"`; in theory, we could use an `aria-live="polite"` instead, but use (I think?) `setTimeout` to delay when it's added in by a second or two so it comes after (a) focus has been managed and (b) the saving message comes to an end.

**Comment by Mottie** [comment]
Nothing has been done since last week 

- Option A would involve changes to the profile code... I haven't looked deeply into this.
- Option B might be easier to implement.

We need to discuss this

**Comment by joshkimux** [comment]
Update on this, I've replicated this again in NoD on JAWs 😢 
@Mottie I'd love to talk about this more and figure out what realistic next steps we might be able to explore

**Comment by davidakennedy** [comment]
I retested this in all the major screen reader combinations, focusing on Chrome/JAWS since that's the last remaining trouble spot. Some notes:

- The focus works correctly in all instances.
- All screen readers announce the alert after an item is saved.
- JAWS announces it as "Email address updated; Main region; Email address updated; Contact information, Heading level 3" I added puncuation there to denote pauses in the speech.
- I suspect that JAWS is announcing correctly because of the recent design system updates to v3 components. The included screenshot shows that the alerts are a v3 component.

![Alert showing that the home phone number is updated and a version 3 component from the VA Design System.](https://github.com/department-of-veterans-affairs/va.gov-team/assets/1473618/15b0c4fb-48f4-4f7f-9d19-f3135f5305b0)

All that said, we can close this ticket. The alerts work as expected. cc: @saderagsdale and @Mottie for awarness.
